docs/hacking.md

```diff
@@ -0,0 +1,176 @@
+> once in the sections below.
+
+### Parsing
+
+The parsing phase translates `.carp` source files into abstract syntax trees
+(AST). In carp, AST nodes are represented using an abstract data type called
+`XObj`. `XObj`s are ubiquitous across the compiler and are used in several
+different phases and contexts. Every `XObj` consists of:
+
+- An `Obj` which is the representation of some carp source code as an abstract
+  data type
+- `Info`: which contains additional information about the source code that
+  generated an `Obj` (e.g. its location in a source file)
+- `Ty`: An valid carp Type for the `Obj`, as determined by the [type
+  system](#type-system).
+
+The following sources are important for parsing:
+
+- `Parsing.hs` -- parsing logic that translates carp source code into abstract
+  syntax.
+- `XObj.hs` -- defines the valid Carp AST nodes.
+
+### Dynamic Evaluator
+
+As stated in the [Macro guide](Macros.md#inner-workings) the dynamic evaluator
+is the central component in the compiler. As the name suggests, the evaluator
+evaluates parsed carp code (`XObjs`) and prepares it for
+[emission](#code-emission). Evaluation entails:
+
+- Expanding macros and dynamic functions
+- Resolving bindings to other forms
+- Requesting type inference for forms
+- Requesting borrow checking for forms
+
+In addition to the `XObjs` corresponding to the source file being compiled, the
+evaluator relies on a `Context`--`Context` is a global object that contains
+state for the compiler. The compiler's `Context` is comprised of several
+environments, defined by the `Env` type--which hold references to known
+bindings. Different environments are used by different phases of the compiler
+to evaluate forms, resolve types, and, generally speaking prepare code for
+emission.
+
+`Binders` are another important abstract data type used in evaluation. Any
+value that's bound to a name in a source program is translated into a `binder`,
+which is comprised of the `XObj` of the form bound to the name, as well as
+additional metadata for the binding. `Binders` are added to the environments in
+the `Context`.
+
+The following sources are important for evaluation:
+
+- `src/Eval.hs` -- this is the entry point for the evaluator.
+- `src/Obj.hs` -- Defines `Context` which carries compiler state across
+  evaluation passes in the form of `Env`s, defines `Env` which holds `Binders`
+from names to `XObjs`.
+- `src/Primitives.hs` -- builtin functions or "keywords" that **do not**
+  evaluate their arguments
+- `src/Commands.hs` -- builtin functions or "keywords" that evaluate their
+  arguments
+- `src/StartingEnv.hs` -- defines the starting environment for the compiler to
+  work with. All commands and primitives are registered here, so that
+evaluation passes can use them.
+- `Lookup.hs` -- Functions for looking up `Binders` in a given environment
+  (`Env`).
+- `Expand.hs` -- Functions for expanding macro forms.
```

Comment:
Actually, this is kind of historical, as this module doesn’t really expand macros any more.

Reply:
Since you're the expert on this part of the system now, maybe you can add some clarifications regarding this – would be good for the future!